### PR TITLE
Issue 104, support dynamic properties in $filter, $orderby and $select

### PR DESCRIPTION
**Pull request by connected-web-development**
Added support for dynamic properties in $select, $filter and $orderby. When the dynamic property is not available, their value is set to 'null' which is then used in ordering for example. If unwanted, the filter option can be used to filter those results just as one would normally do.

### CODE REVIEW

File: OData/test/System.Web.OData.Test/OData/OpenEntityTypeTest.cs [hunk 1]
```diff
@@ -47,6 +48,102 @@ public void Get_OpenEntityType()
...
+            request.Headers.Accept.Add(MediaTypeWithQualityHeaderValue.Parse("application/json;odata.metadata=full"));
+            HttpResponseMessage response = client.SendAsync(request).Result;
+
+            // Assert
+            Assert.True(response.IsSuccessStatusCode);
+            JObject result = JObject.Parse(response.Content.ReadAsStringAsync().Result);
+            var resultArray = result["value"] as JArray;
+            Assert.Equal(6, resultArray.Count);
+            Assert.NotNull(resultArray[2]["Token"]);//customer 2 has a token
+            Assert.NotNull(resultArray[4]["Token"]);//customer 3 has a token
```
Comment (VikingsFan) [inline OData/test/System.Web.OData.Test/OData/OpenEntityTypeTest.cs]: customer 4?

File: OData/src/System.Web.OData/OData/Query/OrderByOpenPropertyNode.cs [hunk 2]
```diff
@@ -0,0 +1,46 @@
+﻿using System.Web.Http;
```
Comment (VikingsFan) [inline OData/src/System.Web.OData/OData/Query/OrderByOpenPropertyNode.cs]: please add Copyright like other files too

File: OData/test/System.Web.OData.Test/OData/OpenEntityTypeTest.cs [hunk 3]
```diff
@@ -213,6 +316,29 @@ private static IList<SimpleOpenCustomer> CreateCustomers()
             return customers;
         }
 
+        [EnableQuery]
+        public IQueryable<SimpleOpenCustomer> Get(ODataQueryOptions<SimpleOpenCustomer> options)
+        {
+            var queryable = CreateCustomers().AsQueryable();
+            //if (options.Filter != null)
```
Comment (VikingsFan) [inline OData/test/System.Web.OData.Test/OData/OpenEntityTypeTest.cs]: remove these lines if it's not comment

**Comment by connected-web-development** [comment]
Comments processed.

**Comment by MrTomWhite** [comment]
@connected-web-development It looks like this produces a merge conflict. You should fix this so the PR can be merged. :package:

**Comment by connected-web-development** [comment]
Merge done

**Comment by VikingsFan** [comment]
@connected-web-development Thank you for responding so quick, the reason this change is not merge yet is we have two features have some relations with your change, for example the untyped scenario, so after we finish these two features (hope will done today), than we can merge your change or you can merge yourself, thanks :) .

File: OData/src/System.Web.OData/OData/Query/OrderByOpenPropertyNode.cs [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+namespace System.Web.OData.Query
+{
+    /// <summary>
+    /// Represents ordering on a dynamic property
+    /// </summary>
+    public class OrderByOpenPropertyNode : OrderByNode
+    {
+        /// <summary>
+        /// Default constructor for a dynamic property order by
```
Comment (VikingsFan) [inline OData/src/System.Web.OData/OData/Query/OrderByOpenPropertyNode.cs]: Initializes a new instance of the <see cref="OrderByOpenPropertyNode"/> class.

File: OData/src/System.Web.OData/OData/Query/OrderByOpenPropertyNode.cs [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+        /// </summary>
+        /// <param name="orderByClause">The order by clause for this open property</param>
```
Comment (VikingsFan) [inline OData/src/System.Web.OData/OData/Query/OrderByOpenPropertyNode.cs]: end with '.'

File: OData/src/System.Web.OData/OData/Query/OrderByOpenPropertyNode.cs [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+        public OrderByOpenPropertyNode(OrderByClause orderByClause)
```
Comment (VikingsFan) [inline OData/src/System.Web.OData/OData/Query/OrderByOpenPropertyNode.cs]: why not call :base(...Direction)

File: OData/src/System.Web.OData/OData/Query/OrderByOpenPropertyNode.cs [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+            }
+
+            OrderByClause = orderByClause;
+            Direction = orderByClause.Direction;
+
+
+            var openPropertyExpression = orderByClause.Expression as SingleValueOpenPropertyAccessNode;
+            if (openPropertyExpression == null)
+            {
+                throw new ODataException(SRResources.OrderByClauseNotSupported);
```
Comment (VikingsFan) [inline OData/src/System.Web.OData/OData/Query/OrderByOpenPropertyNode.cs]: add the the orderby node name?

File: OData/src/System.Web.OData/OData/Query/OrderByOpenPropertyNode.cs [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+            }
+            PropertyName = openPropertyExpression.Name;
+        }
+
+        /// <summary>
+        /// The order by clause
+        /// </summary>
+        public OrderByClause OrderByClause { get; set; }
```
Comment (VikingsFan) [inline OData/src/System.Web.OData/OData/Query/OrderByOpenPropertyNode.cs]: private set?

File: OData/src/System.Web.OData/OData/Query/OrderByNode.cs [hunk 9]
```diff
@@ -43,6 +43,13 @@ public static IList<OrderByNode> CreateCollection(OrderByClause orderByClause)
                 if (clause.Expression is NonentityRangeVariableReferenceNode || clause.Expression is EntityRangeVariableReferenceNode)
                 {
                     result.Add(new OrderByItNode(clause.Direction));
+                    continue;
+                }
+
+                var openPropertyExpression = clause.Expression as SingleValueOpenPropertyAccessNode;
+                if (openPropertyExpression != null)
```
Comment (VikingsFan) [inline OData/src/System.Web.OData/OData/Query/OrderByNode.cs]: change to " is SingleValueOpenPropertyAccessNode "

File: OData/src/System.Web.OData/OData/Query/OrderByQueryOption.cs [hunk 10]
```diff
@@ -245,6 +236,25 @@ private IOrderedQueryable ApplyToCore(IQueryable query, ODataQuerySettings query
...
+                    }
+                    propertiesSoFar.Add(openPropertyNode.PropertyName);
+
+                    if (openPropertyNode.OrderByClause != null)
+                    {
+                        querySoFar = AddOrderByQueryForProperty(query, querySettings, openPropertyNode.OrderByClause, querySoFar, openPropertyNode.Direction, alreadyOrdered);
+                    }
+                    else
+                    {
+                        throw new NotImplementedException(string.Format("Error ordering the dynamic property {0} on type {1}", openPropertyNode.PropertyName, Context.ElementClrType.FullName));
```
Comment (VikingsFan) [inline OData/src/System.Web.OData/OData/Query/OrderByQueryOption.cs]: Please add it into resource string, which will be translate into multiple language

File: OData/src/System.Web.OData/OData/Query/Expressions/FilterBinder.cs [hunk 11]
```diff
@@ -32,6 +32,8 @@ internal class FilterBinder
     {
         private const string ODataItParameterName = "$it";
 
+        private static readonly string _dictionaryStringObjectIndexerName = typeof (Dictionary<string, object>).GetDefaultMembers()[0].Name;
```
Comment (VikingsFan) [inline OData/src/System.Web.OData/OData/Query/Expressions/FilterBinder.cs]: remove white space after "typeof"

please run "build.cmd" to make sure build with any error

File: OData/src/System.Web.OData/OData/Query/Expressions/FilterBinder.cs [hunk 12]
```diff
@@ -233,6 +237,57 @@ private Expression Bind(QueryNode node)
             }
         }
 
+        private Expression BindOpenPropertyAccessQueryNode(SingleValueOpenPropertyAccessNode openNode)
```
Comment (VikingsFan) [inline OData/src/System.Web.OData/OData/Query/Expressions/FilterBinder.cs]: change to BindDynamicPropertyAccessQueryNode(..)

File: OData/src/System.Web.OData/OData/Query/Expressions/FilterBinder.cs [hunk 13]
```diff
@@ -233,6 +237,57 @@ private Expression Bind(QueryNode node)
...
+        {
+            var prop = GetOpenTypeProperty(openNode);
+            var propertyAccessExpression = BindPropertyAccessExpression(openNode, prop);
```
Comment (VikingsFan) [inline OData/src/System.Web.OData/OData/Query/Expressions/FilterBinder.cs]: there is a same function defined as MakeFunctionCall(..), please consider to re-use the existing function.

File: OData/src/System.Web.OData/OData/Query/Expressions/FilterBinder.cs [hunk 14]
```diff
@@ -233,6 +237,57 @@ private Expression Bind(QueryNode node)
...
+                IsNullable(source.Type) && source != _lambdaParameters[ODataItParameterName])
+                propertyAccessExpression = Expression.Property(RemoveInnerNullPropagation(source), prop.Name);
+            else
+            {
+                propertyAccessExpression = Expression.Property(source, prop.Name);
+            }
+            return propertyAccessExpression;
+        }
+
+        private PropertyInfo GetOpenTypeProperty(SingleValueOpenPropertyAccessNode openNode)
```
Comment (VikingsFan) [inline OData/src/System.Web.OData/OData/Query/Expressions/FilterBinder.cs]: make change it as "GetDynamicPropertyDictionary" or "GetDynamicPropertyContainer()"
or make some comment for this function

File: OData/src/System.Web.OData/OData/Query/Expressions/FilterBinder.cs [hunk 15]
```diff
@@ -233,6 +237,57 @@ private Expression Bind(QueryNode node)
...
+        {
+            IEdmStructuredType edmEntityType;
```
Comment (VikingsFan) [inline OData/src/System.Web.OData/OData/Query/Expressions/FilterBinder.cs]: edmStructuredType?

File: OData/src/System.Web.OData/OData/Query/Expressions/FilterBinder.cs [hunk 16]
```diff
@@ -233,6 +237,57 @@ private Expression Bind(QueryNode node)
...
+            {
+                edmEntityType = edmTypeReference.AsEntity().EntityDefinition();
+            }
+            else if (edmTypeReference.IsComplex())
+            {
+                edmEntityType = edmTypeReference.AsComplex().ComplexDefinition();
+            }
+            else
+            {
+                throw Error.NotSupported(SRResources.QueryNodeBindingNotSupported, openNode.Kind, typeof (FilterBinder).Name);
```
Comment (VikingsFan) [inline OData/src/System.Web.OData/OData/Query/Expressions/FilterBinder.cs]: remove the white space after "typeof"

File: OData/test/System.Web.OData.Test/OData/OpenEntityTypeTest.cs [hunk 17]
```diff
@@ -47,6 +48,102 @@ public void Get_OpenEntityType()
...
+            Assert.True(response.IsSuccessStatusCode);
+            JObject result = JObject.Parse(response.Content.ReadAsStringAsync().Result);
+            var resultArray = result["value"] as JArray;
+            Assert.Equal(2, resultArray.Count);
+            Assert.Equal(2, resultArray[0]["CustomerId"]);
+            Assert.Equal(4, resultArray[1]["CustomerId"]);
+        }
+
+        [Fact]
+        public void Get_OpenEntityTypeWithOrderbyDescending()
```
Comment (VikingsFan) [inline OData/test/System.Web.OData.Test/OData/OpenEntityTypeTest.cs]: Merge as one test case and use [Theory]

File: OData/test/System.Web.OData.Test/OData/OpenEntityTypeTest.cs [hunk 18]
```diff
@@ -47,6 +48,102 @@ public void Get_OpenEntityType()
...
+            Assert.Equal(2, resultArray.Count);
+            Assert.Equal(4, resultArray[0]["CustomerId"]);
+            Assert.Equal(2, resultArray[1]["CustomerId"]);
+        }
+
+        [Fact]
+        public void Get_OpenEntityTypeWithFilter()
+        {
+            // Arrange
+            const string RequestUri = "http://localhost/odata/SimpleOpenCustomers?$filter=Token ne null";
```
Comment (VikingsFan) [inline OData/test/System.Web.OData.Test/OData/OpenEntityTypeTest.cs]: test in above, so use other condition

File: OData/src/System.Web.OData/OData/Query/OrderByOpenPropertyNode.cs (hunk 4 above)
Comment (connected-web-development) [inline OData/src/System.Web.OData/OData/Query/OrderByOpenPropertyNode.cs]: I will change the comment though it's hardly informing that a constructor initializes a new instance

**Comment by connected-web-development** [comment]
The only thing that's not fixed is the comment about the orderby ndoe name. However this would require changes in the orderbyproperty node as well as it works the same so the error should be the same. But that would also require the tests to be updated accordingly etc.

**Comment by VikingsFan** [comment]
Checked in 3f4363320308bcf78c19c5d8ad8777034552111d

**Pull request closed without merging**